Add a distribution object to hold quantities with uncertainties

eteq opened the pull request:
This is a WIP that represents combined work of myself, @astrofrog, @mhvk, and @adrn.  The core idea is to create an object that acts like a `Quantity`, *but* instead of being a single value, is a set of samples from a *distribution*.  Any arbitrary operations applied to that object then act on the set of samples, and the result is essentially "automatic" uncertainty propogation (although at the cost of n_samples times the memory usage).

Some items left to do (*possibly* doable by 3.0, but I don't want to jinx it by setting the milestone just yet...):
- [x]  Get all the tests passing (part of the point of creating this now is to ensure there aren't bigger problems on the various additional versions on CI)
- [x]  Documentation on this feature (including statistical caveats - esp when "not enough samples" is a problem)
- [x] Decide on the name: @adrn wants to call it `MarginalDistribution` instead of `Distribution` (I still favor `Distribution`, but @adrn  can weigh in).

Longer-term additions (i.e. would be separate PRs)
* A `JointDistribution` class that carries a set of distribution and allows e.g. easy corner plots, expectation values, etc (@adrn's idea)
* Integrate this with @mhvk's work adapting standard uncertainty propagation following the normal approximation and automatic differentiation (#3715 I believe - that is *complementary* to this, not necessarily a replacement).

astrofrog commented:
There was a bug in the bot, it should not have been closed without the warning so re-opening.

mhvk commented:
Yes, this is still of interest. @eteq - With `erfa` now implemented as ufuncs, more things should work out of the box. It might be worth seeing where it stands.

astrofrog commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,391 @@
+        The variance of this distribution. Shape must match and unit must be
+        compatible with ``center``, or be None (if ``std`` or ``ivar`` are set).
+    var : `Quantity` or None
+        The inversev ariance of this distribution. Shape must match and unit
```
inversev ariance -> inverse variance

astrofrog commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,391 @@
+    var : `Quantity` or None
+        The variance of this distribution. Shape must match and unit must be
+        compatible with ``center``, or be None (if ``std`` or ``ivar`` are set).
+    var : `Quantity` or None
```
var -> ivar

astrofrog commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,391 @@
+        are set).
+    var : `Quantity` or None
+        The variance of this distribution. Shape must match and unit must be
+        compatible with ``center``, or be None (if ``std`` or ``ivar`` are set).
```
Put single backticks around ``None``

astrofrog replied:
(there are other places where this should be done too)

astrofrog commented on docs/units/distribution_stretchgoal_api.rst:
```diff
@@ -0,0 +1,57 @@
+>>> distr12_mul.pdf_
+<Quantity 19.79716280988112 kpc>
+>>> distr12_plus.pdf_std  # analytically should be the quadrature sum of the *pecentage* uncertainty of the components
+<Quantity 4.196109517371489 kpc>
```
We need to be careful here that we still add the underlying samples and not re-generate random numbers, in order to take into account correlated values. For instance, doing ``(d1 + d2)/d1`` should be exactly equivalent to ``1 + d2 / d1``. In general I'm not a fan of mixing the analytical with the Monte-Carlo stuff for that reason.

For instance, should ``pdf_std`` ever be the input analytical value, or should it *always* be the actual ``std`` of the underlying sample?

eteq replied:
As currently implemented everything is Monte Carlo, so this shouldn't present an issue.  In the commits I just pushed up I said explcitly that it's all sampled but that we reserve the *possibility* of analytic representations in the future - so kicking that can down the road to be solved only if necessary.

astrofrog commented:
Ah yes, I'd be fine with ``stats`` too.

eteq commented:
Alright, finally carved out time to do some of the touch-ups on this PR, and re-opened. I think it's *almost* there, so still targeting 3.1.

@astrofrog, I think I addressed all your comments above (except the one that I *think* requires no further action - feel free to comment inline if you disagree).

That leaves two remaining "big" questions (both easy to address once we have an answer):

1. Where this should live. I think I still prefer `units` because it's very closely tied conceptually with `Quantity`.  My perspective here is mainly from "how users will find it" vs "where the code most naturally sits", though.  My concern with `stats` is that this is *not* really a "statistical function/tool" - rather an underlying data object that one can *do* statistics on. But it also doesn't seem like it deserves its own package any more than `quantity` does.  I can be convinced, though, if @mhvk, @astrofrog, and/or @adrn (or anyone else) continue to prefer one of the other choices over `units`.

2. The name of the class - `Distribution` vs `MarginalDistribution`.  I don't see any further comment from @adrn here so I left it as `Distribution`... I do think that's better because it's easier to understand especially given `JointDistribution` does not currently exist (and may never...).  I guess I'm giving @adrn a final chance to given opinion here :wink:

More minor things that are working but could use some feedback: 
1. might be useful for @astrofrog, @mhvk or @adrn to have a look over the doc section I added in the last set of commits. Basically it's a re-working of the API to include narrative description of the content.

2. @mhvk, you might want to look at what I did here to get indexing to work properly.  It works, but there *may* be a more elegant solution using `array_finalize`?

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+Distribution class and associated machinery.
+"""
+
+from __future__ import (absolute_import, division, print_function,
```
Python2...

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+    Parameters
+    ----------
+    distr : array-like
+        The distribution, with sampling along the *leading* axis.
```
Should be `*trailing*`

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+    ----------
+    distr : array-like
+        The distribution, with sampling along the *leading* axis.
+    distr_center : `None` or array-like
```
I would do away with this argument. I see little benefit yet very clear risk of confusion.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        The "center" of this distribution.  It must have the same shape as
+        ``distr``, aside from the (initial) sample dimension. If `None`, the
+        *median* of ``distr``  along the sampling axis will be used.
+    unit : astropy unit
```
And remove the `unit` argument as well - people should just pass in a `Quantity` if they want units - as is, the code allows a non-unit array, and there doesn't seem anything wrong with that.

(Reminded of this because of the comments in `__new__` below.)

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        if item == 'samples':
+            return super(Distribution, result).view(self._distr_cls)
+        else:
+            return Distribution.__new__(self.__class__, result['samples'])
```
Why is this necessary?  I would think you can just `return result` here - it should already be a `Distribution` with the correct unit (if this is to redefine `center`, then all the more reason to get rid of it!!)

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        else:
+            return perc
+
+    def hist(self, maxtoshow=10, **kwargs):
```
For this initial go, I'd remove this or at least make it a function rather than a method. To me it feels too arbitrary to be generally useful, and before you know it, there is a whole new API for setting graph properties attached to the class.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        The standard deviation/σ of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``var`` or ``ivar``
+        are set).
+    var : `Quantity` or `None`
```
I feel there is too much choice here, just stick with `std` which is most common and what goes into `randr` anyway.

Separately, I wondered about adding a `seed` argument, or alternatively a `RandomState` instance. -- although I guess that can be done with the context anyway.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+    Remaining keywords are passed into the `Distribution` constructor
+    """
+    def __new__(cls, poissonval, n_samples=1000, **kwargs):
+        if hasattr(poissonval, 'unit') and 'unit' not in kwargs:
```
Not sure I like allowing any unit except for counts (the only one that makes sense). For now, I would not even allow any unit.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        randshape = np.broadcast(std, center).shape + (n_samples,)
+        distr = center[..., np.newaxis] + np.random.randn(*randshape) * std[..., np.newaxis]
+        self = super(NormalDistribution, cls).__new__(cls, distr, **kwargs)
+        self.distr_std = std
```
I'd get rid of `distr_std` together with `distr_center`.  Note that given that this a subclass of `Distribution`, I think any operation on it will keep the class - even though the distribution may no longer be normal.  Have a look at `Latitude` and `Longitude` on how to avoid that.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        return hists
+
+
+class NormalDistribution(Distribution):
```
Upon further inspection, I'd recommend to remove these classes, or alternatively make them functions that produce `Distribution` instances. See in-line comments below.

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+    # Test that we can initialize directly from a Numpy array if we provide a unit
+    rates = np.array([1, 5, 30, 400])[:, np.newaxis]
+    parr = np.random.poisson(rates, (4, 1000))
+    u.Distribution(parr, unit=u.kpc)
```
Why would this fail without a unit? Also, why have an unphysical example in the test? (Poisson is for counts, i.e., unitless).

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+def test_numpy_init_T():
+    rates = np.array([1, 5, 30, 400])
+    parr = np.random.poisson(rates, (1000, 4))
+    u.Distribution(parr.T, unit=u.kpc)
```
As above, why a unit? Also, this doesn't test the result is correct.

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+def test_quantity_init():
+    # Test that we can initialize directly from a Quantity
+    pq = np.random.poisson(np.array([1, 5, 30, 400])[:, np.newaxis],
+                           (4, 1000)) * u.kpc
```
Sorry to belabor, but why the unit here...

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+        # Number of samples
+        assert self.distr.n_samples == 10000
+
+    def test_n_distr(self):
```
Not sure what this is meant to test!

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+
+    def test_add_quantity(self):
+        distrplus = self.distr + [2000, 0, 0, 500] * u.pc
+        expected = (np.median(self.data, axis=-1) + np.array([2, 0, 0, 0.5])) * self.distr.unit
```
Probably good to check the actual values rather than just the mean.

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+
+
+def test_helper_poisson_samples():
+    centerqadu = [1, 5, 30, 400] * u.adu
```
This unit I can quibble less about, though even so one would normally divide by a gain to make this actual counts... Bottom line, I would stick with unitless for now.

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+
+def test_helper_poisson_exact():
+    pytest.skip('distribution stretch goal not yet implemented')
+    # TODO: what does Poisson u.Distribution mean for a quantity in e.g. kpc?
```
Hey, a comment agreeing with me!

mhvk commented on astropy/units/tests/test_distribution.py:
```diff
@@ -0,0 +1,296 @@
+
+
+def test_helper_normal_exact():
+    pytest.skip('distribution stretch goal not yet implemented')
```
Without being sure about what the desired behaviour should be, I'd delete these "stretch goals".

mhvk commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+
+The core object for this feature is the |distribution|.  Currently, all
+such distributions are Monte Carlo sampled.  While this means each distribution
+may take more memory, it allows arbitrarily complex operations to be performed
```
may take -> takes

mhvk commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+
+  >>> u.Distribution(np.random.poisson(12, (1000))) # doctest: +ELLIPSIS
+  ndarrayDistribution([..., 12,...]) with n_samples=1000
+  >>> pq = np.random.poisson([1, 5, 30, 400], (1000, 4)).T * u.kpc # note the transpose, required to get the sampling on the *last* axis
```
Use a normal distribution here so the `kpc` makes sense.

mhvk commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+Additionally, Poisson and uniform |distribution| classes exist:
+
+  >>> p_dist = u.PoissonDistribution(centerq)
+  >>> uwidth = [10, 20, 10, 55]*u.pc
```
`u.pc` -> `u.count` if you insist on a unit...

mhvk commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+Using Distributions
+===================
+
+This object now acts much like a |quantity| for all but the non-sampled
```
quantity *or ndarray*

mhvk commented:
As for the location, I'd be in favour of a new `uncertainty` module. Mostly because the `Distribution` works with any `ndarray` subclass so is not tied to `Quantity`, so why should it be under `units`? But also because I used this for my `Variable` class attempt, and that one as well as possible extensions mean that likely there will be quite a bit more code.

I think `stats` right now is more of a collection of tools. As such, it does not seem very logical. On the other hand, arguably `Distribution` changes it to become much more or a core module.

astrofrog commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+techniques.
+
+The core object for this feature is the |distribution|.  Currently, all
+such distributions are Monte Carlo sampled.  While this means each distribution
```
Here and maybe elsewhere, hyphenate Monte-Carlo

eteq replied:
I disagree.  In almost all stats literature I'm familiar with you don't hyphenate it.  As a (admittedly slightly flimsy but still useful) source: https://en.wikipedia.org/wiki/Monte_Carlo_method

That said... this *might* be a British vs American English thing?  Maybe we need a cross-Atlantic astropy style guide :wink:

astrofrog commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+on distributions while maintaining their correlation structure. Some specific
+well-behaved distributions (e.g., `~astropy.units.NormalDistribution`) have
+analytic forms which may eventually enable a more compact/efficient
+representation, but this is not currently implemented.
```
Just a note about this - if we ever do this we'll need to make sure we still deal with correlated samples.

mhvk replied:
My `Variable` class does do that. I really should try to resurrect it; it is at a level at least equal to `Distribution`, with the only uncertainy about how far to continue to propagate dependencies.

eteq replied:
@mhvk - I think the thought was to do exactly that once we've had some API experience with this. I guess this may provide a place to do that! (I still don't like the name `Variable` though :wink:)

mhvk replied:
Indeed. And the name of course is open for discussion!

astrofrog commented on docs/units/distributions.rst:
```diff
@@ -0,0 +1,243 @@
+======================
+
+The most conceptually straightforward way to create a distribution is to use an
+array or quantity that carries the samples in the *last* dimension:
```
I feel like it would actually be better to first show the more common case of the well defined distributions and *then* to show how do create arbitrary distributions. Creating the more common ones is easier since it doesn't rely on having to make samples manually. The first example could use only e.g. gaussians and show a simple calculation with a real example (e.g. take an equation such as 1/2 M v^2 and then show how to do uncertainty propagation.

eteq replied:
I see your point here... and that might be more natural with the classmethod/function way

astrofrog commented:
@mhvk: 

>  I also think your documentation already shows that the NormalDistribution etc classes are not needed: one can trivially generate those samples oneself.

I personally disagree with this - to me one of the benefits of this is actually to hide away the actual sampling, and I'd actually see myself using *only* the convenience classes. If I have to create the samples manually anyway then I don't really see the point of having ``Distribution`` since I can just do, in the common case where I'm just dealing with scalars (aside from the samples distribution):

```
x = np.random.normal(0, 1, 1000) * u.m
y = np.random.normal(0.2, 0.3, 1000) * u.s
z = x / y
print(np.mean(z))
```

I guess we are approaching this with different philosophies :)

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        new_dtype = np.dtype({'names': ['samples'],
+                              'formats': [(distr.dtype, (distr.shape[-1],))]})
+        distr_cls = type(distr)
+        if not issubclass(distr_cls, Distribution):
```
The idea is that the next bit of code (which I added) creates a new joint class (so `ndarrayDistribution`, `QuantityDistribution`, etc.). Since that needs the class anyway, this seemed a slightly faster way to do it.

mhvk commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        if not issubclass(distr_cls, Distribution):
+            # Probably should have a dict on the class with these, so
+            # we don't create new classes needlessly.
+            new_name = distr_cls.__name__ + cls.__name__
```
Yes, that is the idea.  Essentially, `Distribution` is a mixin-class that gets added.

mhvk commented:
@astrofrog - I see your point too...  Since my main objection is that in the present design these are separate subclasses which are (but should not be) propagated, how about instead defining functions that create those distributions? Or alternatively and perhaps more easily discovered, make them `Distribution` class methods?

astrofrog commented:
Having thought about it a bit also in the mean time, I really like the idea of the functions. Also if this is in its own subpackage, one could do e.g.:

```
from astropy import distributions as dist
x = dist.normal(0, 1)
y = dist.unoform(0, 1)
```

which is much more concise.

mhvk commented:
Yes, though I'd have it as `from astropy.uncertainty import distributions as dist`... (or just have it be in the top level of `uncertainty` - preferring the module to be a bit more about the concept than the implementation, just as `Quantity` is part of `units`)

bsipocz commented:
Actually I slightly prefer this to be in stats and to move that submodule from the current random dumping of tools toward a more conceptual structure.

mhvk commented:
@bsipocz - must admit I've wondered about whether parts of `stats` shouldn't be affiliated - see long discussion on planet-transit type periodogrammes - so :+1: to bringing more structure to it.

On the flip side, the goal here really is error propagation and that is a nicely self-contained issue which perhaps deserves its own contained sub-module. Indeed, one of my hesitations for having this in `units` is that it would break `units` being a nice self-contained unit, one that perhaps some day might just be transferred to `numpy` or `scipy` (since it is not astronomy specific). One could argue that, similarly, `Distribution` (and my rather poorly named `Variable` class) is something that might eventually make it upstream.

Given this, I'd still prefer a new stand-alone module.

bsipocz commented:
@mhvk - I'm trying to see this from a user's eye (which I can't really do at this point any more, so this is all hypothetically naive): I'm not sure whether I would even look for separate `distributions` module when there is a stats one, the example by Tom above is more intuitive to me if it's all coming from stats. 

```
from astropy.stats import distributions as dist
x = dist.normal(0, 1)
y = dist.uniform(0, 1)
```

Either way, wherever it lands in the namespace, this is a super great addition for 3.1.

mhvk commented:
Possibly... Though in the end users, if they are anything like me (oops), either see example code somewhere or look at the documentation. If it is its own entry under `Data structures and transformations`, e.g., `Uncertainties and Distributions (astropy.uncertainty)`, it would be quite visible.

Indeed, this is perhaps another argument against putting it under `units` - one wouldn't look for it with an entry `Units and Quantities` (obviously, one could rename to `Units, Quantities, Uncertainties and Distributions`...)

Though  `Astrostatistics Tools` is also OK: I do think one would likely look there too, especially if we rename that `Statistical distributions and tools`.

eteq commented:
To quick hit the big-picture items before I can get to addressing the detailed comments:

> the NormalDistribution etc classes are not needed

I'm liking @astrofrog's suggestion of functions over classes.  They basically are just glorified `__init__` methods, which is a sign that they should have been that in the first place! I had imagined they'd do analytic stuff, but I see your point that if they aren't doing that I see@mhvk's point that we should leave the door open for a new class in the future.

I'm a bit partial to classmethods over functions, though. I might try that but if there are objections that's trivial to switch back to regular functions.  It also depends somewhat on the resolution to the "where this lives" question.

> remove the center and unit keywords

I'm not sure I full get what you mean here, @mhvk .  `unit` seems critical if we expect these to be a drop-in replacement for `Quantity`.  Or do you mean that it should always be implied essentially by the presence or absence of units on the input?  I see that point if that's what you mean and could easily enough do that.

But `center` is even more critical for use as a "drop-in" - the idea is that  anyone using the distribution as a drop-in for `Quantity` still wants to know what the value "would have been". Philisophically I can see how that might be good for the concept of a `Distributions`... but I think 

> As for the location

I can accept I'm out-voted on `units` (and I see that it is less relevant now that it also accepts arrays, not just quantities), but am stuck on the `stats` vs `uncertainities`/`distribution` sub-package.  I see the argument for `stats` *but* I am concerned that we may never be able to make this into the "core" `stats` object given that the ship has sailed on a lot of the APIs in there.  But maybe this gives us something to aim for?

(I am somewhat uncharacteristically on the fence on the location question, so am very open to persuading, @astrofrog, @mhvk, @bsipocz, @adrn, or @crawfordsm as stats maintainer)

astrofrog commented:
> But center is even more critical for use as a "drop-in" - the idea is that anyone using the distribution as a drop-in for Quantity still wants to know what the value "would have been". Philisophically I can see how that might be good for the concept of a Distributions...

Just curious, what about if I make a function that generates a bimodal distribution - what should ``center`` be? Isn't that quite specific to single peaked-distributions?

mhvk commented:
Agreed with @astrofrog. Adding to that: withoutt propagating the centre, there seems to be even less point, i.e., a user cannot count on it being there (while a simple work-around for someone who does want to track it is to make one of the samples "special"). Anyway, main reasoning is to start simple. 

On units: my logic is indeed that the user can just put in a `Quantity`. Since an `ndarray` is also an option, the unit is superfluous.

On how to do normal, etc.: class methods would make sense here, my only real worry is a proliferation of them.

adrn commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        A unit of the same sort that `~astropy.units.Quantity` accepts.
+
+    Additional argumnets are passed into the `~astropy.units.Quantity`
+    constructor.
```
Also: `argumnets` -> `arguments`

adrn commented:
I have many thoughts related to things we've discussed in person, but I'll start with the most controversial: I still worry a bit about naming and how to make the API useful for both simple error propagation demos and research cases -- I can provide a few specific use cases of my own if that would be helpful. So, given that, I would actually prefer to see this put into a light-weight affiliated package first. We can even release what you have with 3.1 to get people to try it out, but I worry that we might later want to restructure or change the API, and so it feels premature to put it in the core package so soon.

On naming: My main issue with calling it `Distribution` is that, to me, the name alone implies that it can handle arbitrary N-dimensional probability distributions, but that isn't the case for individual objects. For example, in this implementation, when you pass in multi-dimensional samples, `mean`, `std`, `median`, `percentiles` are actually treating each dimension independently and returning statistics computed on the marginal distributions. (Well, the `mean` is ok I guess...) I worry a bit about mis-use here. So, at absolute minimum, this needs to be stated very clearly and very early in the docs and docstring.

mhvk commented:
I agree with @eteq that this has been on the to-do list for long enough that it should go in astropy proper, though indeed with a clear warning that it is experimental. I'd hope to have it joined by my other work as well, and it would be good to try the interface. I also agree with his reasoning for the name - at least, I use the name `Distribution` similarly.

On the location, perhaps indeed the exploratory aspect is best dealt with in a separate module. How about using `uncertainty` (as in my PR)? It does seem somewhat distinct from the types of statistics tools that we now have in `stats`, being specifically targeted at describing and propagating uncertainties.

eteq commented:
I'm coming around to `uncertainty` on the basis of @mhvk's arguments here, so will try to do that move in the next iteration.

eteq commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+        Parameters
+        ----------
+        maxtoshow : int or `None`
+            The maximum number of distribution elements to show.  If `None`,
```
Hmm, I wanted to avoid phrasing things so that the non-sampled dimensions are "distributions"... I will try to clarify.

eteq commented on astropy/units/distribution.py:
```diff
@@ -0,0 +1,427 @@
+
+    Parameters
+    ----------
+    poissonval : `Quantity`
```
I hate `lambda` as a paremeter because literature is not consistent on this, but `center`... might work even if we remove center elsewhere

eteq commented:
Alright, I think the latest round of commit now rolls together the results of the above discussion.  So this is now in a new `uncertainty` sub-package (which has a "the API might change warning" and corresponding low stability level in the stability table).

Obviously depends on the tests passing... but do you think this is now on target, @astrofrg and @mhvk?

mhvk commented on astropy/uncertainty/tests/test_distribution.py:
```diff
@@ -0,0 +1,305 @@
+
+
+def test_helper_poisson_samples():
+    centerqadu = [1, 5, 30, 400] * u.adu
```
Sorry for being anal, but perhaps use `u.count`? (Which also suggests that both `count` and `electron` should be explicitly allowed as units for poisson - thought I'd quite happily forego units for now...)

mhvk commented on astropy/uncertainty/tests/test_distribution.py:
```diff
@@ -0,0 +1,305 @@
+
+def test_helper_poisson_exact():
+    pytest.skip('distribution stretch goal not yet implemented')
+    # TODO: what does Poisson Distribution mean for a quantity in e.g. kpc?
```
I think this is easy to implement: just don't allow anything but dimensionless for now - better relax later!

mhvk commented on astropy/uncertainty/tests/test_distribution.py:
```diff
@@ -0,0 +1,305 @@
+                                 "for dimensionless quantities")
+
+
+def test_arithmetic_exact():
```
Just delete the test - it is not even correct as written!

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+Introduction
+============
+
+In addition to |quantity|, astropy provides a |distribution| object to represent
```
Delete "in addition to |quantity|"

mhvk commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+__all__ = ['normal', 'poisson', 'uniform', 'uniform_center_width']
+
+
+def normal(center, std=None, var=None, ivar=None, n_samples=1000,
```
I'm actually not sure about having defaults for `std` and `n_samples` here - I think both should be given by the user.

astrofrog replied:
I think we should have:

```
def normal(center, *, std=None, etc.)
```

to make sure the std etc arguments are keyword-only.

mhvk commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+    return distr
+
+
+def poisson(center, n_samples=1000, dcls=Distribution, **kwargs):
```
See above for `n_samples`

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+
+Now we create two |distribution| objects to represent our distributions::
+
+  >>> a = unc.normal(1*u.kpc, std=30*u.pc)
```
I think even if there is a default for `n_samples` (which I think there should not be), we should give an explicit `n_samples` here.

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+operations where error analysis becomes untenable, |distribution| still powers
+through:
+
+  >>> d = unc.poisson(3*u.kpc)
```
Noooooooooooooooooooooooo

eteq replied:
In my defense, I did not just do this blindly: I was considering a case where you actually have binned (i.e., discrete) data binned in kpc.  But I see your point that that is *not* a good intro example.

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+
+  >>> unc.Distribution(np.random.poisson(12, (1000)))  # doctest: +ELLIPSIS
+  ndarrayDistribution([..., 12,...]) with n_samples=1000
+  >>> pq = np.random.poisson([1, 5, 30, 400], (1000, 4)).T * u.kpc # note the transpose, required to get the sampling on the *last* axis
```
Please use something else than `kpc`... `count`?

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+
+Additionally, Poisson and uniform |distribution| creation functions exist:
+
+  >>> unc.poisson(centerq) # doctest: +FLOAT_CMP
```
Same comment...

mhvk commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+Using Distributions
+-------------------
+
+This object now acts much like a |quantity| for all but the non-sampled
```
quantity or ndarray

astrofrog commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+
+def uniform_center_width(center, width, **kwargs):
+    """
+    Create a uniform distribution from lower/upper bounds (instead of center
```
For a normal distribution, we have one function with three options (sigma, fwhm, var). Why not do the same for uniform and have keyword-only arguments, so e.g:

```
def uniform(*, center=None, width=None, min=None, max=None)
```

I don't feel strongly about this, but it could be a nice way to always be explicit.

astrofrog commented on astropy/uncertainty/core.py:
```diff
@@ -0,0 +1,261 @@
+
+    # we set this by hand because the symbolic expression (below) requires scipy
+    # _smad_scale_factor = 1 / scipy.stats.norm.ppf(0.75)
+    _smad_scale_factor = 1.48260221850560203193936104071326553821563720703125
```
Agree

eteq replied:
At the time I think I was thinking other scale factors might be of use to sub-classs... but I see the point that as expressed that really would be more confusing than just multiplying.  So will do.

astrofrog commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+on distributions while maintaining their correlation structure. Some specific
+well-behaved distributions (e.g., the Normal distribution) have
+analytic forms which may eventually enable a more compact/efficient
+representation, but this is not currently implemented.
```
We might want to at least mention how this relates to the uncertainty stuff in the nddata package, and point out that we might somehow combine these in future?

eteq replied:
I added a few words that amount to "this might happen"

astrofrog commented on docs/uncertainty/index.rst:
```diff
@@ -0,0 +1,353 @@
+
+To demonstrate a simple use case for distributions, consider the problem of
+uncertainty propogation of normal distributions.  Assume there are two
+measurements you wish to add.  We start with some initial imports/setup::
```
```suggestion
measurements you wish to add, each with normal uncertainties.  We start
with some initial imports/setup::
```

eteq commented on astropy/uncertainty/__init__.py:
```diff
@@ -0,0 +1,12 @@
+
+
+from .core import *
+from .builtin_distrs import *
```
I was going to suggest `distribution_functions`... but then realize that *also* is an overloaded term!  So I threw up my hands and will go with `distributions` 😜

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+Built-in distribution-creation functions.
+"""
+
+from __future__ import (absolute_import, division, print_function,
```
whoops!  Shows how long this has been brewing, though!

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def normal(center, std=None, var=None, ivar=None, n_samples=1000,
+           dcls=Distribution, **kwargs):
```
I disagree here.  I have some use cases I've been experimenting on in my own science where I would want to make distribution sub-classes (which do not have kwargs at this point but the fact that I had that thought suggests others might).  And given the big red "warning" I think it's good that we are encouraging experimentation here.

I see the point on `cls` vs `dcls`, though, will change.

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+        The standard deviation/σ of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``var`` or ``ivar``
+        are set).
+    var : `~astropy.units.Quantity` or `None`
```
Because it's a convenience function, and lots of users prefer to carry around ivar instead of std, for example. I agree it's rather trivial, but given that it's already a "convenience" function I prefer to make it extra-convenient.

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+    randshape = np.broadcast(std, center).shape + (n_samples,)
+    samples = center[..., np.newaxis] + np.random.randn(*randshape) * std[..., np.newaxis]
+    distr = dcls(samples, **kwargs)
+    distr.distr_std = std
```
Hmm, it was a left-over from before and I felt like it shouldn't be removed.  But on the less-is-more perspective, I see your point, will remove.

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+    """
+    # we convert to arrays because np.random.poisson has trouble with quantities
+    has_unit = False
+    if hasattr(center, 'unit'):
```
I see your point that the example shouldn't use it since kpc is not a counting unit... but I think `electron`/`count`/`adu` are important use cases, particularly given that one of the significant applications here would be for CCDs (via `ccdproc`/`nddata`), which is explicitly made to allow such flexible "counting" units that follow poisson statistics.

So I'd rather be *flexible* here at the outset so that we can experiment.  But perhaps a compromise is to have a warning that is raised for anything other than dimensionless or one of our builtin "counting" units?  I'll give that a try and we'll see how it looks.

eteq commented on astropy/uncertainty/builtin_distrs.py:
```diff
@@ -0,0 +1,192 @@
+    distr : ``dcls``, usually `Distribution`
+        The sampled uniform distribution.
+    """
+    lhasu = hasattr(lower, 'unit')
```
Ah, clever!  Yes that works too, will do.

eteq commented on astropy/uncertainty/core.py:
```diff
@@ -0,0 +1,261 @@
+
+import numpy as np
+
+from .. import visualization
```
I found it really useful to have a quick `.hist()` available when testing this for some science cases.  I admit it's really only useful for the 1D/scalar case though, I admit. Maybe a compromise would be to specifically *only* allow that?

eteq commented on astropy/uncertainty/core.py:
```diff
@@ -0,0 +1,261 @@
+        ----------
+        perc : float or array of floats
+            The desired  precentiles of the distribution (i.e., on [0,100]).
+        kwargs
```
Yeah, I'm thinking we can add a `pdf_quantile` when it is.

eteq commented on astropy/uncertainty/core.py:
```diff
@@ -0,0 +1,261 @@
+
+        Parameters
+        ----------
+        perc : float or array of floats
```
I like the `percent` option - that's a useful end-run around percentile's annoyances.  But I don't want to have our own "quantile" implementation since `numpy` now has one - would rather wait until we can use the numpy one and have *that* be `pdf_quantile`.

eteq commented on astropy/uncertainty/tests/test_distribution.py:
```diff
@@ -0,0 +1,305 @@
+        assert self.distr.n_samples == 10000
+
+    def test_n_distr(self):
+        # Shape of the PDF (note, this is actually the number of values regardless of samples, needs a better name?)
```
Oops, a leftover from the API

eteq commented:
@mhvk and @astrofrog - I think I've gotten everything accounted for (except one noted below)!  Note the unresolved items above are places where I made a command decision or disagreed, but I think they've all at least been touched on there (and most cases I just agreed without comment and implemented).  Highlights include:

* eliminated `hist` (note originally I planned to make `pdf_histogram` and have `hist` only work for scalar distributions... but when I simplified `hist` that much I realized it was superfluous just as @mhvk said!)
* changed uniform to just one function as @astrofrog suggested
* nsamples is now required
* poisson yields a *warning* if its not countable units, but not an exception (see the discussion inline above)

The one thing I did not yet manage is the change @mhvk rightly advocated of not having poisson kpc exampls in the docs.  It turns out that's non-trivial because part of the point was comparing kpc to pc.  This could instead by `normal` presumably, but then requires a bunch of fiddly checking of numbers that I don't have time left for right now.  We can get that in later though (before release) because it's a doc change!  (Although if @astrofrog or @mhvk have time to do that, feel free to do so and push it up here)

mhvk commented:
sorry, in china and have to run - am happy with doctest done later. would still prefer poisson to only allow specific units (dimensionless, count, electron, *not* ADU, which should be divided by gain to become Poisson).

eteq commented:
OK, I think I've fixed up all the test failures and a few random things. I also managed to convert the docs to be count-based instead of kpc (although it's a bit forced and it might be good to-rephrase if someone has a better way to do this... but that should be a follow on PR as the current does work wither with or without the counts-as-error point @mhvk raises).

I still don't agree with the final comment from @mhvk, so I think the way to proceed is to leave it to @astrofrog to make the deciding vote.  If you agree with @mhvk you should just be able to change the `warn`'s in `poisson` to raise exceptions, and otherwise it should be good to merge as-is.

pllim commented:
This is great! Any plans to also make a tutorial with this functionality?
